internal/notifier/slack.go

```diff
@@ -118,13 +122,56 @@ func (s *Slack) Post(ctx context.Context, event eventv1.Event) error {
 
 	payload.Attachments = []SlackAttachment{a}
 
-	err := postMessage(ctx, s.URL, s.ProxyURL, s.CertPool, payload, func(request *retryablehttp.Request) {
-		if s.Token != "" {
-			request.Header.Add("Authorization", "Bearer "+s.Token)
-		}
-	})
-	if err != nil {
+	opts := []postOption{
+		withRequestModifier(func(request *retryablehttp.Request) {
+			if s.Token != "" {
+				request.Header.Add("Authorization", "Bearer "+s.Token)
+			}
+		}),
+	}
+	if s.ProxyURL != "" {
+		opts = append(opts, withProxy(s.ProxyURL))
+	}
+	if s.CertPool != nil {
+		opts = append(opts, withCertPool(s.CertPool))
+	}
+	if s.URL == "https://slack.com/api/chat.postMessage" {
+		opts = append(opts, withResponseValidator(s.validateResponse))
+	}
+
+	if err := postMessage(ctx, s.URL, payload, opts...); err != nil {
 		return fmt.Errorf("postMessage failed: %w", err)
 	}
+
 	return nil
 }
+
+// validateResponse validates that a chat.postMessage API response is successful.
+// chat.postMessage API always returns 200 OK.
+// See https://api.slack.com/methods/chat.postMessage.
+//
+// On the other hand, incoming webhooks return more expressive HTTP status codes.
+// See https://api.slack.com/messaging/webhooks#handling_errors.
+func (s *Slack) validateResponse(resp *http.Response) error {
+	// Clone resp.Body so it can be read again in postMessage error handling.
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("unable to read response body: %w", err)
+	}
+	resp.Body = io.NopCloser(bytes.NewBuffer(body))
+
+	type slackResponse struct {
+		Ok    bool   `json:"ok"`
+		Error string `json:"error"`
+	}
+	slackResp := slackResponse{}
+	if err := json.Unmarshal(body, &resp); err != nil {
```

Comment:
This defer will close the new body you assigned below, and not the original body which is what you want. You don't need to defer this because there's already a defer outside in case reading errors out. You only need to close the original body reader if the ReadAll() succeeds.

However, since reading the body in the response validator is already starting to get messed up, I think a much better idea is reading the body before calling the response validator and passing just a `[]byte` to the validator instead.

Reply:
> I think a much better idea is reading the body before calling the response validator and passing just a `[]byte` to the validator instead.

Surely that's a better idea! I rewrote the response validator in that way.
Ideally we should read the response body only when it's necessary for e.g. error reporting, but we can do more optimization in another pull request.